Deprecate Remaining Legacy Circuit Library Classes and Update References

AsadullahGalib007 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ X] I have added the tests to cover my changes.
- [ X] I have updated the documentation accordingly.
- [ X] I have read the CONTRIBUTING document.
-->

### Summary
This PR continues the deprecation effort initiated in #13604 and #14478 by addressing additional legacy components within the Qiskit `circuit.library` module. Specifically, it deprecates certain classes and updates the relevant docstrings accordingly.

### Details and Comments
- The following legacy quantum circuits have been deprecated: `ExactReciprocal`, `LinearAmplitudeFunction`, `QuantumVolume`, and `PhaseOracle`.
- Docstrings have been updated for the following: `PiecewisePolynomialPauliRotationsGate`, `zz_feature_map`, and `z_feature_map`.

### Pending Items for Future PRs
- Decide the future of the `GR`, `GRX`, `GRY`, and `GRZ` circuits.
- Add associated tests for the gates/functions.

### Related Issue
This PR resolves #14478.

AsadullahGalib007 committed:
Depricated the Quantum Circuits #14478_1

AsadullahGalib007 committed:
Updated the docstrings #14478_3

AsadullahGalib007 committed:
Fixed minor typos #14478

AsadullahGalib007 committed:
Fixed minor typos #14478

Cryoris commented:
Thanks for tackling this @AsadullahGalib007. We do have an existing workflow for deprecations of this kind, you don't have to add manual deprecation warnings, nor manual Sphinx blocks. 

See e.g. here how it should be done: 
https://github.com/Qiskit/qiskit/blob/6161b56216a11622a5f962015ddbe8c140d76592/qiskit/circuit/library/n_local/pauli_two_design.py#L172
The deprecation version should be Qiskit 2.2 and this pull request would also need a release note that mentions the deprecated class and their replacements.

AsadullahGalib007 commented:
> Thanks for tackling this @AsadullahGalib007. We do have an existing workflow for deprecations of this kind, you don't have to add manual deprecation warnings, nor manual Sphinx blocks.
> 
> See e.g. here how it should be done:
> 
> https://github.com/Qiskit/qiskit/blob/6161b56216a11622a5f962015ddbe8c140d76592/qiskit/circuit/library/n_local/pauli_two_design.py#L172
> 
> 
> The deprecation version should be Qiskit 2.2 and this pull request would also need a release note that mentions the deprecated class and their replacements.

Thank you for bringing this to my attention. I will update the snippets as soon as possible.

AsadullahGalib007 committed:
Updated deprecated warnings under @deprecate_func and the docstrings

AsadullahGalib007 commented:
Hi @Cryoris, 
I've removed the manual warnings and Sphinx blocks and applied the @deprecate_func decorator as per the existing workflow. I've also added a release note mentioning the deprecated class and its replacement. Additionally, I updated the docstrings for `PiecewisePolynomialPauliRotationsGate`, `z_feature_map`, and `zz_feature_map`. Please let me know if any further changes are needed!

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -222,7 +223,8 @@ def z_feature_map(
         >>> ry = TwoLocal(3, "ry", "cz", reps=1).decompose()
-        >>> classifier = z_feature_map(3, reps=1) + ry
+        >>> classifier = z_feature_map(3, reps=1)
+        >>> classifier.append(ry, list(range(classifier.num_qubits)))
```
This is a good catch! Would you mind updating the snippet to use the new `n_local` function instead of the `TwoLocal` class?

AsadullahGalib007 replied:
Sure! I will do it asap.

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -277,7 +279,7 @@ def zz_feature_map(
     Examples:
 
-        >>> from qiskit.circuit.library import ZZFeatureMap
+        >>> from qiskit.circuit.library.data_preparation import zz_feature_map
```
We encourage high-level imports (the ones that are actually API documented) so it would be better to leave this as:
```suggestion
        >>> from qiskit.circuit.library import zz_feature_map
```

Cryoris commented on qiskit/circuit/library/data_preparation/pauli_feature_map.py:
```diff
@@ -289,7 +291,7 @@ def zz_feature_map(
         >>> classifier = zz_feature_map(3).compose(efficient_su2(3))
         >>> classifier.num_parameters
-        15
+        27
```
Nice catch!

AsadullahGalib007 replied:
Thanks!

AsadullahGalib007 committed:
Update qiskit/circuit/library/data_preparation/pauli_feature_map.py

High level imports

Co-authored-by: Julien Gacon <gaconju@gmail.com>

AsadullahGalib007 committed:
high-level import for z_feature_map

AsadullahGalib007 committed:
Example of n_local

AsadullahGalib007 commented:
Hi @Cryoris,
I’ve made the changes you suggested, including the use of `n_local` function instead of `TwoLocal`. Let me know if anything else needs adjusting!

Cryoris commented:
This looks good, it only needs a release note listing the deprecated circuits and their replacements. See e.g. https://github.com/Qiskit/qiskit/blob/308df7ac22ff91d9f49fbdb6998578dbcda60ba9/releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml#L19 (you don't have to explain anything in detail, just list the gates).

We'll also hold off merging until the 2.1 release, since this is only for 2.2.

AsadullahGalib007 committed:
Added release note for deprecated circuit classes

AsadullahGalib007 commented:
> This looks good, it only needs a release note listing the deprecated circuits and their replacements. See e.g.
> 
> https://github.com/Qiskit/qiskit/blob/308df7ac22ff91d9f49fbdb6998578dbcda60ba9/releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml#L19
> 
> (you don't have to explain anything in detail, just list the gates).
> We'll also hold off merging until the 2.1 release, since this is only for 2.2.

Thank you for the feedback.

I've added the release note listing the deprecated circuits and their replacements following the format. Please let me know if any further changes are required.

P.S. I’m currently participating in UnitaryHack 2025 and working on completing this issue. If everything looks good, I would appreciate it if you could assign the issue to me and mark it as closed.

Cryoris commented on releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml:
```diff
@@ -27,3 +27,8 @@ deprecations_circuits:
       * :class:`.MCXGrayCode` ``-->``  :func:`.synth_mcx_gray_code`
       * :class:`.MCXRecursive` ``-->``  :func:`.synth_mcx_n_dirty_i15`
       * :class:`.MCXVChain` ``-->`` :func:`.synth_mcx_n_clean_m15`
+      * :class:`.ExactReciprocal` ``-->`` :class:`.ExactReciprocalGate`
```
You'll need to put these in a separate release note, since your PR will only be included in the 2.2 release, but this release note will be included in 2.1 🙂

AsadullahGalib007 replied:
My bad! Just to confirm — that means I should create a folder for 2.1 inside the release notes and place a YAML file there with the deprecated items. Is that correct?

alexanderivrii replied:
Hi @AsadullahGalib007, please take a look at the section "Release notes" in the Qiskit contributing guide: https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#release-notes.

Cryoris commented:
I can assign you on the issue if you leave a comment there (otherwise I can't assign you) 🙂 The issue is going to be closed once this PR merges, which is likely going to be on Friday.

AsadullahGalib007 committed:
Added a release note

AsadullahGalib007 committed:
Fixed the linting-error

AsadullahGalib007 commented:
Can anyone assist me with the `qiskit._accelerate.circuit_library`? I'm facing the following error: `ImportError: cannot import name '_accelerate' from partially initialized module 'qiskit'` when running tox.

Cryoris commented:
Did you install qiskit in editable mode (meaning via `pip install -e .`)? If yes, then you have to recompile the Rust library when it changed, which could happen when you updated from `main`. Try running
```
python setup.py build_rust --inplace --release
```
to see if it fixes things 🙂

1ucian0 commented on releasenotes/notes/deprecate-circuit-libraries-d39877bc0e962f05.yaml:
Are the release note complete? For example, should `quantum_volume` be there?

AsadullahGalib007 replied:
There is one more release note for deprecated items. `quantum_volume` is present there: [releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml](https://github.com/Qiskit/qiskit/blob/308df7ac22ff91d9f49fbdb6998578dbcda60ba9/releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml#L25)

AsadullahGalib007 replied:
Updated

AsadullahGalib007 committed:
updated tests to work with latest gate-functions

AsadullahGalib007 commented:
> Did you install qiskit in editable mode (meaning via `pip install -e .`)? If yes, then you have to recompile the Rust library when it changed, which could happen when you updated from `main`. Try running
> 
> ```
> python setup.py build_rust --inplace --release
> ```
> 
> to see if it fixes things 🙂

Thank you @Cryoris, for your clear instructions. I have run all the tests on my end, and they have passed. I have added a release note and updated the tests to work with the latest gate/functions. This should fully address the issue. Kindly review and share your thoughts.

1ucian0 commented on test/python/circuit/library/test_quantum_volume.py:
```diff
@@ -22,18 +22,24 @@
+        with self.assertWarns(DeprecationWarning):
+            left = QuantumVolume(4, 4, seed=2024, flatten=True)
+        with self.assertWarns(DeprecationWarning):
+            right = QuantumVolume(4, 4, seed=2024, flatten=True)
```
It is perfectly fine to use the same context for both calls:

```suggestion
        with self.assertWarns(DeprecationWarning):
            left = QuantumVolume(4, 4, seed=2024, flatten=True)
            right = QuantumVolume(4, 4, seed=2024, flatten=True)
```

AsadullahGalib007 committed:
py312 OK - all tests passed

Cryoris commented on releasenotes/notes/deprecate-circuit-libraries-d39877bc0e962f05.yaml:
```diff
@@ -0,0 +1,9 @@
+    The following classes in the circuit library are deprecated as of Qiskit 2.2
+    and will be removed in Qiskit 3.0. They have been replaced with modern gate equivalents.
+
+      * :class:`.ExactReciprocal` ``-->`` :class:`.ExactReciprocalGate`
```
This misses `QuantumVolume` in the deprecations

AsadullahGalib007 replied:
There is one more release note for deprecated items. `quantum_volume` is present there:
https://github.com/Qiskit/qiskit/blob/308df7ac22ff91d9f49fbdb6998578dbcda60ba9/releasenotes/notes/deprecate-library-circuits-4f8b234500f90d50.yaml#L25

Should I include one more time?

Cryoris replied:
Yes 🙂 Since your releasenote lists the circuits that are deprecated, it should list quantum volume again. The note you were linking just contains a list of examples of users should use the circuits, but it's not an official list of what is deprecated.

Cryoris commented on test/python/circuit/library/test_phase_and_bitflip_oracles.py:
```diff
@@ -33,76 +33,126 @@ class TestPhaseOracleAndGate(QiskitTestCase):
-            with self.subTest(use_gate=use_gate):
-                self.assertListEqual(expected_valid, result_valid)
-                self.assertListEqual(expected_invalid, result_invalid)
+    def test_phase_oracle_gate_variable_order(self, expression, var_order, good_states):
```
It would be nice to merge these tests. You could, for example, add a bool argument to decide whether the gate or the circuit is used, as is done in
https://github.com/Qiskit/qiskit/blob/09bd717a86bca13680e81b41fc27b47cb73cb685/test/python/circuit/library/test_quadratic_form.py#L74
or you could just have a for loop in the function itself, something along the lines of
```python
@data(...)
@unpack
def test_phase_oracle_gate_variable_order(self, expression, var_order, good_states):
    for use_gate in [True, False]:
        if use_gate:
            oracle = PhaseOracleGate(....)
        else:
            with self.assertWarns(DeprecationWarning):
                oracle = PhaseOracle(...)
```
This way we can avoid duplicating all the tests.

AsadullahGalib007 replied:
Right. I will update it asap.

AsadullahGalib007 committed:
addressed reviewer comments on release-note and tests

1ucian0 commented on test/python/circuit/library/test_quantum_volume.py:
```diff
@@ -22,18 +22,24 @@
+        with self.assertWarns(DeprecationWarning):
+            left = QuantumVolume(4, 4, seed=28, classical_permutation=False)
+        with self.assertWarns(DeprecationWarning):
+            right = QuantumVolume(4, 4, seed=28, classical_permutation=False)
```
```suggestion
        with self.assertWarns(DeprecationWarning):
            left = QuantumVolume(4, 4, seed=28, classical_permutation=False)
            right = QuantumVolume(4, 4, seed=28, classical_permutation=False)
```

1ucian0 commented on test/python/circuit/library/test_quantum_volume.py:
```diff
@@ -22,18 +22,24 @@
+        with self.assertWarns(DeprecationWarning):
+            left = QuantumVolume(4, 4, seed=3, classical_permutation=True)
+        with self.assertWarns(DeprecationWarning):
+            right = QuantumVolume(4, 4, seed=3, classical_permutation=True)
```
```suggestion
        with self.assertWarns(DeprecationWarning):
            left = QuantumVolume(4, 4, seed=3, classical_permutation=True)
            right = QuantumVolume(4, 4, seed=3, classical_permutation=True)
```

AsadullahGalib007 committed:
Update test/python/circuit/library/test_quantum_volume.py

reviewer suggestion

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

AsadullahGalib007 committed:
Update test/python/circuit/library/test_quantum_volume.py

reviewer suggestion

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

AsadullahGalib007 committed:
Update test/python/circuit/library/test_quantum_volume.py

reviewer suggestion

Co-authored-by: Luciano Bello <766693+1ucian0@users.noreply.github.com>

1ucian0 reviewed (approved):
LGTM

Thanks a lot for your contribution! The bounty for #14478 can be fully assigned to you.

AsadullahGalib007 commented:
> LGTM
> 
> Thanks a lot for your contribution! The bounty for #14478 can be fully assigned to you.

Thank you all for your kind guidance and support throughout the process. I'm grateful for the opportunity to contribute and look forward to contributing more in the future.
